### PR TITLE
add parallel mesh decomposer

### PR DESCRIPTION
**Pull request by vmont**

### CODE REVIEW

**Comment by buildbot-princeton** [comment]
The files changed require a buildbot test. Buildbot has started.

**Comment by buildbot-princeton** [comment]
The files changed require a buildbot test. Buildbot has started.

**Comment by buildbot-princeton** [comment]
Your changes failed the buildbot test. See [build 424](http://specfem3d-buildbot.princeton.edu/builders/specfem3d_builder/builds/424). Fix your mistakes and push your changes back to your branch vmont/specfem3d:devel. This pull-request will be reopened automatically.         You don't need to create a new pull request

**Comment by buildbot-princeton** [comment]
Your changes failed the buildbot test. See [build 425](http://specfem3d-buildbot.princeton.edu/builders/specfem3d_builder/builds/425). Fix your mistakes and push your changes back to your branch vmont/specfem3d:devel. This pull-request will be reopened automatically.         You don't need to create a new pull request

**Comment by vmont** [comment]
Hi all,

I got an error on builbot compilation. It  seems to occur in subroutines I didn't modify. 
I add subroutines in src/shared/parallel.f90 and I got an error in subroutine src/shared/shared_par.f90 and in scotch that I didn't modify. 

I also added new files in decompose_mesh.

On my side all going ok. So I didn't find the problem. How can we fix it?  

Thanks.

 @luet @QuLogic

**Comment by komatits** [comment]
Hi Vadim,

Tomorrow afternoon let us meet and fix that together.

Thanks,
Dimitri.

On 02/10/2016 12:46 PM, Vadim Monteiller wrote:
> Hi all,
>
> I got an error on builbot compilation. It seems to occur in subroutines
> I didn't modify.
> I add subroutines in src/shared/parallel.f90 and I got an error in
> subroutine src/shared/shared_par.f90 and in scotch that I didn't modify.
>
> I also added new files in decompose_mesh.
>
> On my side all going ok. So I didn't find the problem. How can we fix it?
>
> Thanks.
>
> @luet <https://github.com/luet> @QuLogic <https://github.com/QuLogic>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/geodynamics/specfem3d/pull/694#issuecomment-182327979>.
>

-- 
Dimitri Komatitsch
CNRS Research Director (DR CNRS), Laboratory of Mechanics and Acoustics,
UPR 7051, Marseille, France    http://komatitsch.free.fr

**Comment by buildbot-princeton** [comment]
The files changed require a buildbot test. Buildbot has started.

**Comment by buildbot-princeton** [comment]
Your changes failed the buildbot test. See [build 426](http://specfem3d-buildbot.princeton.edu/builders/specfem3d_builder/builds/426). Fix your mistakes and push your changes back to your branch vmont/specfem3d:devel. This pull-request will be reopened automatically.         You don't need to create a new pull request

**Comment by buildbot-princeton** [comment]
The files changed require a buildbot test. Buildbot has started.

**Comment by buildbot-princeton** [comment]
Your changes failed the buildbot test. See [build 427](http://specfem3d-buildbot.princeton.edu/builders/specfem3d_builder/builds/427). Fix your mistakes and push your changes back to your branch vmont/specfem3d:devel. This pull-request will be reopened automatically.         You don't need to create a new pull request

**Comment by buildbot-princeton** [comment]
The files changed require a buildbot test. Buildbot has started.

**Comment by buildbot-princeton** [comment]
Your changes failed the buildbot test. See [build 428](http://specfem3d-buildbot.princeton.edu/builders/specfem3d_builder/builds/428). Fix your mistakes and push your changes back to your branch vmont/specfem3d:devel. This pull-request will be reopened automatically.         You don't need to create a new pull request

**Comment by buildbot-princeton** [comment]
The files changed require a buildbot test. Buildbot has started.

**Comment by buildbot-princeton** [comment]
Your changes passed the buildbot test. See [build 429](http://specfem3d-buildbot.princeton.edu/builders/specfem3d_builder/builds/429).

**Comment by QuLogic** [comment]
Wait! Sorrry, I was making comments on an outdated commit; there are still a couple of issues though, so please do not merge.

File: Makefile.in [hunk 1]
```diff
@@ -243,6 +243,7 @@ SUBDIRS = \
 # default targets for the pure Fortran version
 DEFAULT = \
 	xdecompose_mesh \
+	xdecompose_mesh_paral \
```
Comment (QuLogic) [inline Makefile.in]: Why not name it `xdecompose_mesh_mpi` like the file name?

File: src/decompose_mesh/rules.mk [hunk 2]
```diff
@@ -106,7 +130,10 @@ endif
 
 $O/decompose_mesh.dec.o: $O/part_decompose_mesh.dec.o $O/fault_scotch.dec.o ${SCOTCH_DIR}/include/scotchf.h $O/shared_par.shared_module.o $(COND_MPI_OBJECTS)
 $O/program_decompose_mesh.dec.o: $O/decompose_mesh.dec.o $O/shared_par.shared_module.o $(COND_MPI_OBJECTS)
-
+$O/program_decompose_mesh_mpi.mpidec.o: $O/shared_par.shared_module.o $(COND_MPI_OBJECTS)
```
Comment (QuLogic) [inline src/decompose_mesh/rules.mk]: There are a few other dependencies missing, I think. From my look through:
*    decompose_mesh/module_mesh.f90
       * fault_scotch
*    decompose_mesh/module_partition.f90
       * module_qsort
*    decompose_mesh/program_decompose_mesh_mpi.f90
       * module_mesh
       * module_database
       * module_partition

Also, it is not necessary to specify `$O/shared_par.shared_module.o` here; it is already a dependency from the `$O/%.dec.o` rule below.

File: src/decompose_mesh/module_database.f90 [hunk 3]
```diff
@@ -0,0 +1,457 @@
...
+
+  integer, private, parameter                 :: IIN_database=49
+  integer, private                            :: ier
+  integer, private, dimension(:), allocatable :: node_loc
+
+contains
+
+!----------------------------------
+!
+!  set upt the arrays for database
```
Comment (QuLogic) [inline src/decompose_mesh/module_database.f90]: upt -> up

File: src/decompose_mesh/module_database.f90 [hunk 4]
```diff
@@ -0,0 +1,457 @@
...
+          node_loc(inode) = glob2loc_nodes(nodes_ibelm(inode, i))
+       end do
+       if (ipart(iE) == myrank +1) write(IIN_database) iE_loc, node_loc(1:NGNOD2D) 
+    end do
+    
+  end subroutine write_my_boundary
+
+!----------------------------------
+!
+!  distributed adjancy table
```
Comment (QuLogic) [inline src/decompose_mesh/module_database.f90]: adjancy -> adjacency

File: src/decompose_mesh/module_database.f90 [hunk 5]
```diff
@@ -0,0 +1,457 @@
...
+          endif
+       enddo
+    end if
+    close(IIN_database)
+    
+  end subroutine write_database
+
+!----------------------------------
+!
+!  count elements in boudary
```
Comment (QuLogic) [inline src/decompose_mesh/module_database.f90]: boudary -> boundary

File: src/decompose_mesh/module_database.f90 [hunk 6]
```diff
@@ -0,0 +1,457 @@
...
+       iE=ibelm(i)
+       if (ipart(iE) == myrank +1) nspec2D_loc = nspec2D_loc + 1
+    end do
+    write(IIN_database) iflag, nspec2D_loc
+
+  end subroutine count_my_boundary
+
+!----------------------------------
+!
+!  write elements in boudary
```
Comment (QuLogic) [inline src/decompose_mesh/module_database.f90]: boudary -> boundary

**Comment by QuLogic** [comment]
OK, sorry if you got comments twice; even though BuildBot passed, I think there is still a dependency issue that may cause problems in parallel builds. Please see my comment in the `rules.mk` file above.

**Comment by komatits** [comment]
accept

On 02/10/2016 04:38 PM, buildbot-princeton wrote:
> Your changes passed the buildbot test. See build 429
> <http://specfem3d-buildbot.princeton.edu/builders/specfem3d_builder/builds/429>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/geodynamics/specfem3d/pull/694#issuecomment-182431440>.
>

-- 
Dimitri Komatitsch
CNRS Research Director (DR CNRS), Laboratory of Mechanics and Acoustics,
UPR 7051, Marseille, France    http://komatitsch.free.fr

File: src/decompose_mesh/module_database.f90 (hunk 6 above)
Comment (komatits) [inline src/decompose_mesh/module_database.f90]: Fixed.

On 02/10/2016 07:29 PM, Elliott Sales de Andrade wrote:
> In src/decompose_mesh/module_database.f90
> <https://github.com/geodynamics/specfem3d/pull/694#discussion_r52501445>:
>
>> +    integer,   dimension(nspec2D),             intent(in) :: ibelm
>> +
>> +    integer ::  i, nspec2D_loc, iE
>> +
>> +    nspec2D_loc = 0
>> +    do i=1, nspec2D
>> +       iE=ibelm(i)
>> +       if (ipart(iE) == myrank +1) nspec2D_loc = nspec2D_loc + 1
>> +    end do
>> +    write(IIN_database) iflag, nspec2D_loc
>> +
>> +  end subroutine count_my_boundary
>> +
>> +!----------------------------------
>> +!
>> +!  write elements in boudary
>
> boudary -> boundary
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/geodynamics/specfem3d/pull/694/files#r52501445>.
>

-- 
Dimitri Komatitsch
CNRS Research Director (DR CNRS), Laboratory of Mechanics and Acoustics,
UPR 7051, Marseille, France    http://komatitsch.free.fr

File: src/decompose_mesh/module_database.f90 (hunk 4 above)
Comment (komatits) [inline src/decompose_mesh/module_database.f90]: Fixed.

On 02/10/2016 07:29 PM, Elliott Sales de Andrade wrote:
> In src/decompose_mesh/module_database.f90
> <https://github.com/geodynamics/specfem3d/pull/694#discussion_r52501360>:
>
>> +    integer :: i,  iE, iE_loc, inode
>> +
>> +    do i=1, nspec2D
>> +       iE=ibelm(i)
>> +       iE_loc=glob2loc_elmnt(iE)
>> +       do inode = 1, NGNOD2D
>> +          node_loc(inode) = glob2loc_nodes(nodes_ibelm(inode, i))
>> +       end do
>> +       if (ipart(iE) == myrank +1) write(IIN_database) iE_loc, node_loc(1:NGNOD2D)
>> +    end do
>> +
>> +  end subroutine write_my_boundary
>> +
>> +!----------------------------------
>> +!
>> +!  distributed adjancy table
>
> adjancy -> adjacency
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/geodynamics/specfem3d/pull/694/files#r52501360>.
>

-- 
Dimitri Komatitsch
CNRS Research Director (DR CNRS), Laboratory of Mechanics and Acoustics,
UPR 7051, Marseille, France    http://komatitsch.free.fr

File: src/decompose_mesh/module_database.f90 (hunk 3 above)
Comment (komatits) [inline src/decompose_mesh/module_database.f90]: Fixed.

On 02/10/2016 07:28 PM, Elliott Sales de Andrade wrote:
> In src/decompose_mesh/module_database.f90
> <https://github.com/geodynamics/specfem3d/pull/694#discussion_r52501241>:
>
>> +  integer                                     :: size_adjancy
>> +  integer, dimension(:),  allocatable         :: adjcy, id_adjcy
>> +
>> +  integer                                     :: max_elmnts_by_node
>> +  integer, dimension(:),  allocatable         :: nelmnts_by_node
>> +  integer, dimension(:,:),allocatable         :: elmnts_by_node
>> +
>> +  integer, private, parameter                 :: IIN_database=49
>> +  integer, private                            :: ier
>> +  integer, private, dimension(:), allocatable :: node_loc
>> +
>> +contains
>> +
>> +!----------------------------------
>> +!
>> +!  set upt the arrays for database
>
> upt -> up
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/geodynamics/specfem3d/pull/694/files#r52501241>.
>

-- 
Dimitri Komatitsch
CNRS Research Director (DR CNRS), Laboratory of Mechanics and Acoustics,
UPR 7051, Marseille, France    http://komatitsch.free.fr

File: Makefile.in (hunk 1 above)
Comment (komatits) [inline Makefile.in]: Done.

On 02/10/2016 07:24 PM, Elliott Sales de Andrade wrote:
> In Makefile.in
> <https://github.com/geodynamics/specfem3d/pull/694#discussion_r52500644>:
>
>> @@ -243,6 +243,7 @@ SUBDIRS = \
>>  # default targets for the pure Fortran version
>>  DEFAULT = \
>>  	xdecompose_mesh \
>> +	xdecompose_mesh_paral \
>
> Why not name it |xdecompose_mesh_mpi| like the file name?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/geodynamics/specfem3d/pull/694/files#r52500644>.
>

-- 
Dimitri Komatitsch
CNRS Research Director (DR CNRS), Laboratory of Mechanics and Acoustics,
UPR 7051, Marseille, France    http://komatitsch.free.fr

**Comment by komatits** [comment]
Fixed.

On 02/10/2016 07:31 PM, Elliott Sales de Andrade wrote:
> OK, sorry if you got comments twice; even though BuildBot passed, I
> think there is still a dependency issue that may cause problems in
> parallel builds. Please see my comment in the |rules.mk| file above.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/geodynamics/specfem3d/pull/694#issuecomment-182517939>.
>

-- 
Dimitri Komatitsch
CNRS Research Director (DR CNRS), Laboratory of Mechanics and Acoustics,
UPR 7051, Marseille, France    http://komatitsch.free.fr

File: src/decompose_mesh/rules.mk (hunk 2 above)
Comment (komatits) [inline src/decompose_mesh/rules.mk]: Thanks!
Fixed.

Elliott when you have time could you please check our fix?
(we are not very familiar with what to put in rules.mk)
Thanks.


On 02/10/2016 07:26 PM, Elliott Sales de Andrade wrote:
> In src/decompose_mesh/rules.mk
> <https://github.com/geodynamics/specfem3d/pull/694#discussion_r52500943>:
>
>> @@ -106,7 +130,10 @@ endif
>>
>>  $O/decompose_mesh.dec.o: $O/part_decompose_mesh.dec.o $O/fault_scotch.dec.o ${SCOTCH_DIR}/include/scotchf.h $O/shared_par.shared_module.o $(COND_MPI_OBJECTS)
>>  $O/program_decompose_mesh.dec.o: $O/decompose_mesh.dec.o $O/shared_par.shared_module.o $(COND_MPI_OBJECTS)
>> -
>> +$O/program_decompose_mesh_mpi.mpidec.o: $O/shared_par.shared_module.o $(COND_MPI_OBJECTS)
>
> There are a few other dependencies missing, I think. From my look through:
>
>   * decompose_mesh/module_mesh.f90
>       o fault_scotch
>   * decompose_mesh/module_partition.f90
>       o module_qsort
>   * decompose_mesh/program_decompose_mesh_mpi.f90
>       o module_mesh
>       o module_database
>       o module_partition
>
> Also, it is not necessary to specify |$O/shared_par.shared_module.o|
> here; it is already a dependency from the |$O/%.dec.o| rule below.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/geodynamics/specfem3d/pull/694/files#r52500943>.
>

-- 
Dimitri Komatitsch
CNRS Research Director (DR CNRS), Laboratory of Mechanics and Acoustics,
UPR 7051, Marseille, France    http://komatitsch.free.fr

**Pull request merged**